File: crates/lvz-gw-tui/src/md.rs
//! Lightweight terminal markdown: an inline styler (`**bold**`, `*italic*`/`_italic_`, `` `code` ``,
//! `~~strike~~`) plus a width-aware wrapper for styled segments. Hand-rolled (no `pulldown-cmark`) —
//! it only needs the inline subset, applied per line, so block structure (headings, fenced code) is
//! handled by the caller's line classifier.

use ratatui::style::{Color, Modifier, Style};
use unicode_width::UnicodeWidthStr;

/// A run of text with a resolved style.
pub type Segment = (String, Style);

/// The first index `>= from` at which `needle` occurs in `chars` (single char).
fn find(chars: &[char], from: usize, needle: char) -> Option<usize> {
    (from..chars.len()).find(|&i| chars[i] == needle)
}

/// The first index `>= from` at which the two-char `a b` sequence occurs.
fn find2(chars: &[char], from: usize, a: char, b: char) -> Option<usize> {
    (from..chars.len().saturating_sub(1)).find(|&i| chars[i] == a && chars[i + 1] == b)
}

/// Parse one line's inline markdown into styled segments. An unterminated marker is treated as
/// literal text (so stray `*` or backticks never eat the rest of the line).
pub fn inline(line: &str) -> Vec<Segment> {
    let base = Style::default();
    let chars: Vec<char> = line.chars().collect();
    let mut out: Vec<Segment> = Vec::new();
    let mut plain = String::new();
    let mut i = 0;
    macro_rules! flush {
        () => {
            if !plain.is_empty() {
                out.push((std::mem::take(&mut plain), base));
            }
        };
    }
    while i < chars.len() {
        let c = chars[i];
        // `code`
        if c == '`' {
            if let Some(close) = find(&chars, i + 1, '`') {
                flush!();
                out.push((
                    chars[i + 1..close].iter().collect(),
                    Style::default().fg(Color::Cyan),
                ));
                i = close + 1;
                continue;
            }
        }
        // **bold**
        if c == '*' && chars.get(i + 1) == Some(&'*') {
            if let Some(close) = find2(&chars, i + 2, '*', '*') {
                flush!();
                out.push((
                    chars[i + 2..close].iter().collect(),
                    base.add_modifier(Modifier::BOLD),
                ));
                i = close + 2;
                continue;
            }
        }
        // ~~strike~~
        if c == '~' && chars.get(i + 1) == Some(&'~') {
            if let Some(close) = find2(&chars, i + 2, '~', '~') {
                flush!();
                out.push((
                    chars[i + 2..close].iter().collect(),
                    base.add_modifier(Modifier::CROSSED_OUT),
                ));
                i = close + 2;
                continue;
            }
        }
        // *italic* or _italic_
        if (c == '*' || c == '_') && chars.get(i + 1).is_some_and(|n| !n.is_whitespace()) {
            if let Some(close) = find(&chars, i + 1, c) {
                flush!();
                out.push((
                    chars[i + 1..close].iter().collect(),
                    base.add_modifier(Modifier::ITALIC),
                ));
                i = close + 1;
                continue;
            }
        }
        plain.push(c);
        i += 1;
    }
    flush!();
    if out.is_empty() {
        out.push((String::new(), base));
    }
    out
}

/// Greedily wrap styled segments to `width` display columns, splitting a run mid-way when needed and
/// coalescing consecutive same-style characters into one span per row.
pub fn wrap(segments: &[Segment], width: usize) -> Vec<Vec<Segment>> {
    let width = width.max(1);
    let mut rows: Vec<Vec<Segment>> = vec![Vec::new()];
    let mut col = 0usize;
    for (text, style) in segments {
        for ch in text.chars() {
            let cw = UnicodeWidthStr::width(ch.to_string().as_str()).max(1);
            if col + cw > width && col > 0 {
                rows.push(Vec::new());
                col = 0;
            }
            let row = rows.last_mut().expect("at least one row");
            match row.last_mut() {
                Some((s, st)) if *st == *style => s.push(ch),
                _ => row.push((ch.to_string(), *style)),
            }
            col += cw;
        }
    }
    rows
}

// ---------------------------------------------------------------------------------------------
// Tables — a block construct: rows are buffered by the caller, then rendered aligned as a unit.
// ---------------------------------------------------------------------------------------------

/// Column alignment from a table's separator row.
#[derive(Clone, Copy)]
enum Align {
    Left,
    Center,
    Right,
}

/// Display width of a string in terminal columns.
fn disp(s: &str) -> usize {
    UnicodeWidthStr::width(s)
}

/// Whether a line looks like a table row (trimmed, starts with `|`).
pub fn is_table_row(line: &str) -> bool {
    let t = line.trim_start();
    t.starts_with('|') && t.chars().nth(1).is_some()
}

/// Split a `| a | b |` row into trimmed cells, preserving internal empty cells.
fn parse_row(line: &str) -> Vec<String> {
    let t = line.trim();
    let t = t.strip_prefix('|').unwrap_or(t);
    let t = t.strip_suffix('|').unwrap_or(t);
    t.split('|').map(|c| c.trim().to_string()).collect()
}

/// Whether a line is a table *separator* row (`|---|:--:|`): every cell is dashes with optional
/// leading/trailing colons.
fn is_separator_row(line: &str) -> bool {
    let cells = parse_row(line);
    !cells.is_empty()
        && cells.iter().all(|c| {
            let c = c.trim();
            c.contains('-') && c.chars().all(|ch| ch == '-' || ch == ':')
        })
}

/// Per-column alignment parsed from the separator row's colons.
fn parse_aligns(line: &str) -> Vec<Align> {
    parse_row(line)
        .iter()
        .map(|c| {
            let c = c.trim();
            match (c.starts_with(':'), c.ends_with(':')) {
                (true, true) => Align::Center,
                (false, true) => Align::Right,
                _ => Align::Left,
            }
        })
        .collect()
}

/// Truncate `s` to `w` display columns, appending `…` when it doesn't fit.
fn truncate_to_width(s: &str, w: usize) -> String {
    if disp(s) <= w {
        return s.to_string();
    }
    if w == 0 {
        return String::new();
    }
    let mut acc = String::new();
    let mut used = 0;
    for ch in s.chars() {
        let cw = UnicodeWidthStr::width(ch.to_string().as_str()).max(1);
        if used + cw > w - 1 {
            break;
        }
        acc.push(ch);
        used += cw;
    }
    acc.push('…');
    acc
}

/// Pad `s` to `w` display columns per `align` (assumes `s` already fits).
fn pad_to_width(s: &str, w: usize, align: Align) -> String {
    let d = disp(s);
    if d >= w {
        return s.to_string();
    }
    let pad = w - d;
    match align {
        Align::Right => format!("{}{}", " ".repeat(pad), s),
        Align::Left => format!("{}{}", s, " ".repeat(pad)),
        Align::Center => {
            let l = pad / 2;
            format!("{}{}{}", " ".repeat(l), s, " ".repeat(pad - l))
        }
    }
}

/// Shrink column widths (widest first) until the row fits `avail`, so a wide table truncates cells
/// rather than overflowing the terminal.
fn fit_widths(widths: &mut [usize], avail: usize) {
    let seps = 3 * widths.len().saturating_sub(1);
    let budget = avail.saturating_sub(seps).max(widths.len());
    while widths.iter().sum::<usize>() > budget {
        let (idx, max) = widths
            .iter()
            .enumerate()
            .max_by_key(|(_, w)| **w)
            .map(|(i, w)| (i, *w))
            .unwrap();
        if max <= 3 {
            break;
        }
        widths[idx] -= 1;
    }
}

/// The `i`th cell of a row, or "" when the row is short.
fn nth_cell(row: &[String], i: usize) -> &str {
    row.get(i).map(String::as_str).unwrap_or("")
}

/// Render buffered table lines into aligned, styled visual rows (header bold, a `─┼─` rule, then
/// data), fitted to `width`. Returns `None` when the lines aren't a valid table (no separator row),
/// so the caller can fall back to plain rendering.
pub fn render_table(lines: &[String], width: usize) -> Option<Vec<Vec<Segment>>> {
    if lines.len() < 2 || !is_separator_row(&lines[1]) {
        return None;
    }
    let header = parse_row(&lines[0]);
    let aligns = parse_aligns(&lines[1]);
    let data: Vec<Vec<String>> = lines[2..].iter().map(|l| parse_row(l)).collect();
    let ncols = header
        .len()
        .max(data.iter().map(Vec::len).max().unwrap_or(0))
        .max(1);

    let mut widths = vec![1usize; ncols];
    for (i, w) in widths.iter_mut().enumerate() {
        *w = disp(nth_cell(&header, i));
        for row in &data {
            *w = (*w).max(disp(nth_cell(row, i)));
        }
        *w = (*w).max(1);
    }
    fit_widths(&mut widths, width);

    let sep = Style::default().fg(Color::DarkGray);
    let align = |i: usize| *aligns.get(i).unwrap_or(&Align::Left);
    let build = |row: &[String], style: Style| -> Vec<Segment> {
        let mut out = Vec::new();
        for (i, &w) in widths.iter().enumerate() {
            if i > 0 {
                out.push((" │ ".to_string(), sep));
            }
            let fitted = truncate_to_width(nth_cell(row, i), w);
            out.push((pad_to_width(&fitted, w, align(i)), style));
        }
        out
    };

    let mut out = Vec::new();
    out.push(build(
        &header,
        Style::default().add_modifier(Modifier::BOLD),
    ));
    let mut rule = Vec::new();
    for (i, w) in widths.iter().enumerate() {
        if i > 0 {
            rule.push(("─┼─".to_string(), sep));
        }
        rule.push(("─".repeat(*w), sep));
    }
    out.push(rule);
    for row in &data {
        out.push(build(row, Style::default()));
    }
    Some(out)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn inline_styles_bold_italic_and_code() {
        let segs = inline("a **b** c `d` _e_");
        // Reconstruct the text to confirm markers are consumed, not literal.
        let text: String = segs.iter().map(|(t, _)| t.as_str()).collect();
        assert_eq!(text, "a b c d e");
        assert!(segs
            .iter()
            .any(|(t, s)| t == "b" && s.add_modifier.contains(Modifier::BOLD)));
        assert!(segs
            .iter()
            .any(|(t, s)| t == "d" && s.fg == Some(Color::Cyan)));
        assert!(segs
            .iter()
            .any(|(t, s)| t == "e" && s.add_modifier.contains(Modifier::ITALIC)));
    }

    #[test]
    fn unterminated_marker_stays_literal() {
        let segs = inline("2 * 3 = 6");
        let text: String = segs.iter().map(|(t, _)| t.as_str()).collect();
        assert_eq!(text, "2 * 3 = 6");
        assert!(segs
            .iter()
            .all(|(_, s)| !s.add_modifier.contains(Modifier::ITALIC)));
    }

    #[test]
    fn wrap_splits_and_preserves_style() {
        let segs = vec![("abcdef".to_string(), Style::default())];
        let rows = wrap(&segs, 3);
        assert_eq!(rows.len(), 2);
        let r0: String = rows[0].iter().map(|(t, _)| t.as_str()).collect();
        assert_eq!(r0, "abc");
    }

    #[test]
    fn detects_table_and_separator_rows() {
        assert!(is_table_row("| a | b |"));
        assert!(is_table_row("  |x|"));
        assert!(!is_table_row("just text"));
        assert!(is_separator_row("|---|:--:|"));
        assert!(is_separator_row("| :-- | --: |"));
        assert!(!is_separator_row("| a | b |"));
    }

    #[test]
    fn renders_an_aligned_table() {
        let lines = vec![
            "| Name | Score |".to_string(),
            "|------|------:|".to_string(),
            "| alice | 3 |".to_string(),
            "| bob | 100 |".to_string(),
        ];
        let rows = render_table(&lines, 80).expect("valid table");
        // header + rule + 2 data rows.
        assert_eq!(rows.len(), 4);
        // Header cells are bold.
        assert!(rows[0]
            .iter()
            .any(|(t, s)| t.contains("Name") && s.add_modifier.contains(Modifier::BOLD)));
        // The rule row is drawn with box characters.
        let rule: String = rows[1].iter().map(|(t, _)| t.as_str()).collect();
        assert!(rule.contains('─') && rule.contains('┼'));
        // Right-aligned Score column pads on the left ("  3").
        let alice: String = rows[2].iter().map(|(t, _)| t.as_str()).collect();
        assert!(alice.contains("  3"), "right-aligned, got {alice:?}");
    }

    #[test]
    fn non_table_lines_are_rejected() {
        // No separator row ⇒ not a table (caller falls back to plain rendering).
        let lines = vec!["| a | b |".to_string(), "| c | d |".to_string()];
        assert!(render_table(&lines, 80).is_none());
    }

    #[test]
    fn a_too_wide_table_shrinks_to_fit() {
        let lines = vec![
            "| col |".to_string(),
            "|-----|".to_string(),
            format!("| {} |", "x".repeat(200)),
        ];
        let rows = render_table(&lines, 40).unwrap();
        for row in &rows {
            let w: usize = row.iter().map(|(t, _)| disp(t)).sum();
            assert!(w <= 40, "row exceeded width: {w}");
        }
    }
}

File: src/Lavoisier/Gateway/Tui/Md.hs
-- | Lightweight terminal markdown for the TUI: an inline styler (@**bold**@, @*italic*@\/@_italic_@,
-- @\`code\`@, @~~strike~~@), a width-aware wrapper for styled segments, and an aligned table
-- renderer. Hand-rolled — it only needs the inline subset, applied per line, so block structure
-- (headings, fenced code) is handled by the caller's line classifier.
--
-- Ported from Rust @lvz-gw-tui@ @md.rs@. Where the Rust leans on @ratatui@'s 'Style' and the
-- @unicode-width@ crate, the port carries its own minimal 'Style' (rendered to SGR escapes by 'sgr')
-- and its own 'charWidth' — no TUI framework, in keeping with the tree's hand-rolled adapters.
module Lavoisier.Gateway.Tui.Md
  ( -- * Styling
    Color (..),
    Style (..),
    defaultStyle,
    fg,
    bold,
    italic,
    crossedOut,
    sgr,
    resetSgr,

    -- * Widths
    charWidth,
    displayWidth,

    -- * Inline markdown
    Segment,
    inline,
    wrap,

    -- * Tables
    isTableRow,
    isSeparatorRow,
    renderTable,
  )
where

import Data.Char (ord)
import Data.Text (Text)
import Data.Text qualified as T

-- --- style ----------------------------------------------------------------------------------------

-- | The eight-colour subset the TUI uses (plus a dim grey), as ANSI SGR foreground codes.
data Color = Cyan | Green | Yellow | Red | DarkGray
  deriving stock (Eq, Show)

-- | A resolved text style: an optional foreground colour plus the three inline modifiers.
data Style = Style
  { stFg :: Maybe Color,
    stBold :: Bool,
    stItalic :: Bool,
    stCrossedOut :: Bool
  }
  deriving stock (Eq, Show)

-- | Unstyled.
defaultStyle :: Style
defaultStyle = Style Nothing False False False

fg :: Color -> Style -> Style
fg c s = s {stFg = Just c}

bold, italic, crossedOut :: Style -> Style
bold s = s {stBold = True}
italic s = s {stItalic = True}
crossedOut s = s {stCrossedOut = True}

-- | The SGR escape sequence that turns @style@ on. Always paired with 'resetSgr'.
sgr :: Style -> Text
sgr s
  | null codes = ""
  | otherwise = "\ESC[" <> T.intercalate ";" codes <> "m"
  where
    codes =
      concat
        [ ["1" | stBold s],
          ["3" | stItalic s],
          ["9" | stCrossedOut s],
          maybe [] (pure . colorCode) (stFg s)
        ]
    colorCode = \case
      Cyan -> "36"
      Green -> "32"
      Yellow -> "33"
      Red -> "31"
      -- Bright black reads as grey on every common palette; plain 30 is invisible on dark themes.
      DarkGray -> "90"

-- | Turn every attribute back off.
resetSgr :: Text
resetSgr = "\ESC[0m"

-- --- widths ---------------------------------------------------------------------------------------

-- | Terminal columns one character occupies: 0 for combining marks, 2 for the East-Asian wide and
-- fullwidth ranges (CJK, Hangul, kana, fullwidth forms, most emoji), 1 otherwise. A deliberately
-- small stand-in for the @unicode-width@ crate — enough for the text this UI actually renders.
charWidth :: Char -> Int
charWidth c
  | n == 0 = 0
  | inAny combining = 0
  | inAny wide = 2
  | otherwise = 1
  where
    n = ord c
    inAny = any (\(lo, hi) -> n >= lo && n <= hi)
    combining = [(0x0300, 0x036F), (0x1AB0, 0x1AFF), (0x1DC0, 0x1DFF), (0x20D0, 0x20F0), (0xFE00, 0xFE0F), (0xFE20, 0xFE2F)]
    wide =
      [ (0x1100, 0x115F), -- Hangul Jamo
        (0x2E80, 0x303E), -- CJK radicals, Kangxi, CJK symbols
        (0x3041, 0x33FF), -- kana, Hangul compat jamo, CJK compat
        (0x3400, 0x4DBF), -- CJK ext A
        (0x4E00, 0x9FFF), -- CJK unified
        (0xA000, 0xA4CF), -- Yi
        (0xAC00, 0xD7A3), -- Hangul syllables
        (0xF900, 0xFAFF), -- CJK compat ideographs
        (0xFE30, 0xFE6F), -- CJK compat forms
        (0xFF00, 0xFF60), -- fullwidth forms
        (0xFFE0, 0xFFE6),
        (0x1F300, 0x1F64F), -- emoji
        (0x1F900, 0x1F9FF),
        (0x20000, 0x3FFFD) -- CJK ext B+
      ]

-- | Display width of a string in terminal columns.
displayWidth :: Text -> Int
displayWidth = T.foldl' (\acc c -> acc + charWidth c) 0

-- --- inline markdown ------------------------------------------------------------------------------

-- | A run of text with a resolved style.
type Segment = (Text, Style)

-- | Parse one line's inline markdown into styled segments. An unterminated marker is treated as
-- literal text (so a stray @*@ or backtick never eats the rest of the line).
inline :: Text -> [Segment]
inline line = case go (T.unpack line) "" of
  [] -> [("", defaultStyle)]
  segs -> segs
  where
    base = defaultStyle

    flush plain rest = [(T.pack (reverse plain), base) | not (null plain)] <> rest

    go [] plain = flush plain []
    go cs@(c : rest) plain
      -- `code`
      | c == '`',
        Just (body, after) <- untilChar '`' rest =
          flush plain ((T.pack body, fg Cyan base) : go after "")
      -- \**bold**
      | c == '*',
        take 1 rest == "*",
        Just (body, after) <- untilPair '*' '*' (drop 1 rest) =
          flush plain ((T.pack body, bold base) : go after "")
      -- ~~strike~~
      | c == '~',
        take 1 rest == "~",
        Just (body, after) <- untilPair '~' '~' (drop 1 rest) =
          flush plain ((T.pack body, crossedOut base) : go after "")
      -- \*italic* or _italic_
      | c == '*' || c == '_',
        (r0 : _) <- rest,
        not (isSpace' r0),
        Just (body, after) <- untilChar c rest =
          flush plain ((T.pack body, italic base) : go after "")
      | otherwise = go rest (c : plain)
      where
        _unused = cs

    isSpace' ch = ch == ' ' || ch == '\t'

    -- Split at the next occurrence of `needle`, returning (body, remainder-after-needle).
    untilChar needle s = case break (== needle) s of
      (_, []) -> Nothing
      (body, _ : after) -> Just (body, after)

    -- Split at the next `a b` pair, returning (body, remainder-after-pair).
    untilPair a b = search ""
      where
        search _ [] = Nothing
        search acc (x : y : ys)
          | x == a && y == b = Just (reverse acc, ys)
          | otherwise = search (x : acc) (y : ys)
        search _ [_] = Nothing

-- | Greedily wrap styled segments to @width@ display columns, splitting a run mid-way when needed
-- and coalescing consecutive same-style characters into one span per row.
wrap :: [Segment] -> Int -> [[Segment]]
wrap segments width0 =
  let (rows, cur, _) = foldl' step ([], [], 0) [(ch, st) | (t, st) <- segments, ch <- T.unpack t]
   in map (map (\(t, st) -> (T.pack (reverse t), st)) . reverse) (reverse (cur : rows))
  where
    width = max 1 width0
    step (rows, cur, col) (ch, st)
      | cw <- charWidth ch,
        col + max 1 cw > width && col > 0 =
          (cur : rows, [([ch], st)], max 1 (charWidth ch))
      | otherwise = (rows, push ch st cur, col + max 1 (charWidth ch))
    push ch st ((t, s) : more) | s == st = ((ch : t, s) : more)
    push ch st cur = (([ch], st) : cur)

-- --- tables ---------------------------------------------------------------------------------------

-- | Column alignment from a table's separator row.
data Align = AlignLeft | AlignCenter | AlignRight
  deriving stock (Eq, Show)

-- | Whether a line looks like a table row (trimmed, starts with @|@ and has more after it).
isTableRow :: Text -> Bool
isTableRow line = "|" `T.isPrefixOf` t && T.length t > 1
  where
    t = T.stripStart line

-- | Split a @| a | b |@ row into trimmed cells, preserving internal empty cells.
parseRow :: Text -> [Text]
parseRow line = map T.strip (T.splitOn "|" stripped)
  where
    t = T.strip line
    stripped = dropSuffix "|" (dropPrefix "|" t)
    dropPrefix p s = maybe s id (T.stripPrefix p s)
    dropSuffix p s = maybe s id (T.stripSuffix p s)

-- | Whether a line is a table __separator__ row (@|---|:--:|@): every cell is dashes with optional
-- leading\/trailing colons.
isSeparatorRow :: Text -> Bool
isSeparatorRow line =
  not (null cells) && all ok cells
  where
    cells = parseRow line
    ok c = T.any (== '-') c' && T.all (\ch -> ch == '-' || ch == ':') c'
      where
        c' = T.strip c

-- | Per-column alignment parsed from the separator row's colons.
parseAligns :: Text -> [Align]
parseAligns = map alignOf . parseRow
  where
    alignOf cell = case (":" `T.isPrefixOf` c, ":" `T.isSuffixOf` c) of
      (True, True) -> AlignCenter
      (False, True) -> AlignRight
      _ -> AlignLeft
      where
        c = T.strip cell

-- | Truncate to @w@ display columns, appending @…@ when it doesn't fit.
truncateToWidth :: Text -> Int -> Text
truncateToWidth s w
  | displayWidth s <= w = s
  | w == 0 = ""
  | otherwise = go (T.unpack s) "" 0
  where
    go [] acc _ = T.pack (reverse acc) <> "…"
    go (c : cs) acc used
      | used + max 1 (charWidth c) > w - 1 = T.pack (reverse acc) <> "…"
      | otherwise = go cs (c : acc) (used + max 1 (charWidth c))

-- | Pad to @w@ display columns per @align@ (assumes the text already fits).
padToWidth :: Text -> Int -> Align -> Text
padToWidth s w align
  | d >= w = s
  | otherwise = case align of
      AlignRight -> spaces pad <> s
      AlignLeft -> s <> spaces pad
      AlignCenter -> spaces l <> s <> spaces (pad - l)
  where
    d = displayWidth s
    pad = w - d
    l = pad `div` 2
    spaces n = T.replicate n " "

-- | Shrink column widths (widest first) until a row fits @avail@, so a wide table truncates cells
-- rather than overflowing the terminal.
fitWidths :: [Int] -> Int -> [Int]
fitWidths widths avail = go widths
  where
    seps = 3 * max 0 (length widths - 1)
    budget = max (length widths) (max 0 (avail - seps))
    go ws
      | sum ws <= budget = ws
      | mx <= 3 = ws
      | otherwise = go (take i ws <> [mx - 1] <> drop (i + 1) ws)
      where
        mx = maximum ws
        i = length (takeWhile (/= mx) ws)

-- | The @i@th cell of a row, or @\"\"@ when the row is short.
nthCell :: [Text] -> Int -> Text
nthCell row i = case drop i row of
  (c : _) -> c
  [] -> ""

-- | Render buffered table lines into aligned, styled visual rows (header bold, a @─┼─@ rule, then
-- data), fitted to @width@. Returns 'Nothing' when the lines aren't a valid table (no separator
-- row), so the caller can fall back to plain rendering.
renderTable :: [Text] -> Int -> Maybe [[Segment]]
renderTable lines_ width = case lines_ of
  (hdr : sepLine : rest) | isSeparatorRow sepLine -> Just (table hdr sepLine rest)
  _ -> Nothing
  where
    table hdr sepLine rest = build header (bold defaultStyle) : rule : map (`build` defaultStyle) dataRows
      where
        header = parseRow hdr
        aligns = parseAligns sepLine
        dataRows = map parseRow rest
        ncols = max 1 (maximum (length header : map length dataRows))
        natural = [max 1 (maximum (displayWidth (nthCell header i) : map (\r -> displayWidth (nthCell r i)) dataRows)) | i <- [0 .. ncols - 1]]
        widths = fitWidths natural width
        sep = fg DarkGray defaultStyle
        alignAt i = case drop i aligns of (a : _) -> a; [] -> AlignLeft
        build row style =
          concat
            [ [(" │ ", sep) | i > 0] <> [(padToWidth (truncateToWidth (nthCell row i) w) w (alignAt i), style)]
            | (i, w) <- zip [0 ..] widths
            ]
        rule =
          concat
            [ [("─┼─", sep) | i > 0] <> [(T.replicate w "─", sep)]
            | (i, w) <- zip [(0 :: Int) ..] widths
            ]
